Signal/src/views/OWSSystemMessageCell.m

```diff
@@ -0,0 +1,205 @@
+    }
+
+    UIImage *icon = [self iconForInteraction:self.interaction];
+    self.imageView.image = [icon imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
+    self.imageView.tintColor = [self iconColorForInteraction:self.interaction];
+    self.titleLabel.text = [OWSSystemMessageCell titleForInteraction:self.interaction];
+    self.titleLabel.textColor = [self textColorForInteraction:self.interaction];
+
+    [self setNeedsLayout];
+}
+
+- (UIColor *)textColorForInteraction:(TSInteraction *)interaction
+{
+    return [UIColor colorWithRGBHex:0x303030];
+}
+
+- (UIColor *)iconColorForInteraction:(TSInteraction *)interaction
+{
+    // "Phone", "Shield" and "Hourglass" icons have a lot of "ink" so they
+    // are less dark for balance.
+    return [UIColor colorWithRGBHex:0x404040];
+}
+
+- (UIImage *)iconForInteraction:(TSInteraction *)interaction
+{
+    UIImage *result = nil;
+
+    if ([interaction isKindOfClass:[TSErrorMessage class]]) {
+        switch (((TSErrorMessage *)self.interaction).errorType) {
+            case TSErrorMessageInvalidKeyException:
+            case TSErrorMessageNonBlockingIdentityChange:
+            case TSErrorMessageWrongTrustedIdentityKey:
+            case TSErrorMessageMissingKeyId:
+                result = [UIImage imageNamed:@"system_message_security"];
+                break;
+            case TSErrorMessageNoSession:
+            case TSErrorMessageInvalidMessage:
+            case TSErrorMessageDuplicateMessage:
+            case TSErrorMessageInvalidVersion:
+            case TSErrorMessageUnknownContactBlockOffer:
+                result = [UIImage imageNamed:@"system_message_info"];
+                break;
+        }
+    } else if ([interaction isKindOfClass:[TSInfoMessage class]]) {
+        switch (((TSInfoMessage *)self.interaction).messageType) {
+            case TSInfoMessageUserNotRegistered:
+                result = [UIImage imageNamed:@"system_message_info"];
+                break;
+            case TSInfoMessageTypeSessionDidEnd:
+            case TSInfoMessageTypeUnsupportedMessage:
+            case TSInfoMessageAddToContactsOffer:
+                result = [UIImage imageNamed:@"system_message_info"];
+                break;
+            case TSInfoMessageTypeGroupUpdate:
+            case TSInfoMessageTypeGroupQuit:
+                result = [UIImage imageNamed:@"system_message_group"];
+                break;
+            case TSInfoMessageTypeDisappearingMessagesUpdate:
+                result = [UIImage imageNamed:@"system_message_timer"];
+                break;
+        }
+    } else if ([interaction isKindOfClass:[TSCall class]]) {
+        result = [UIImage imageNamed:@"system_message_call"];
+    } else {
+        OWSFail(@"Unknown interaction type");
+        return nil;
+    }
+    OWSAssert(result);
+    return result;
+}
+
++ (NSString *)titleForInteraction:(TSInteraction *)interaction
+{
+    // TODO: Should we move the copy generation into this view?
```

Comment:
I'd like to, but I suspect that a bunch of our logging logic is using that method. And it'll be non-trivial to find all of the cases where that usage is being done since they overloaded `[NSObject description]`.  I avoid overloading `[NSObject description]`.